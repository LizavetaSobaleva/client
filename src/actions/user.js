import axios from "axios"
import { setUser } from '../reducers/userReducer'


export const instanceAxios = axios.create({
     baseURL: process.env.REACT_APP_API_URL + "/api/"
})

export const registration = async (name, email, password, dispatch) => {
   try {
        const response = await instanceAxios.post('auth/registration', {
            name,
            email,
            password
        })
        alert(response.data.message)
        dispatch(login(email, password))
   } catch (e) {
        alert(e.response.data.message)
   }
}

export const login = (email, password) => {
     return async dispatch => {
          try {
               const response = await instanceAxios.post('auth/login', {
                   email,
                   password
               })
               dispatch(setUser(response.data.user))
               localStorage.setItem('token', response.data.token)
          } catch (e) {
               alert(e.response.data.message)
          }
     }
}

export const auth = () => {
     return async dispatch => {
          try {
               const response = await instanceAxios.get('auth/auth',
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               localStorage.setItem('token', response.data.token)
               dispatch(setUser(response.data.user))
          } catch (e) {
               console.log(e.response.data.message)
               localStorage.removeItem('token')
          }
     }
}

export const uploadAvatar = (file) => {
     return async dispatch => {
          try {
               const formData = new FormData()
               formData.append('file', file)
               const response = await instanceAxios.post('files/avatar', formData,
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               dispatch(setUser(response.data))
          } catch (e) {
               console.log(e)
          }
     }
}

export const deleteAvatar = () => {
     return async dispatch => {
          try {
               const response = await instanceAxios.delete('files/avatar',
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               dispatch(setUser(response.data))
          } catch (e) {
               console.log(e)
          }
     }
}