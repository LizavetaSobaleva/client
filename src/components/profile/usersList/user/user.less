@import '../../../../utils/colors.less';

@import '../../../../utils/colors.less';

.user {
    padding: 8px 12px;
    border-bottom: 1px solid @lightgrey;
    display: grid;
    grid-template-columns: 1.5fr 2fr 2fr 2fr 0.5fr;
    color: @subtext;
    font-weight: 400;
    align-items: center;
    transition: ease-in 50ms all;

    &:hover {
        background-color: #f9fafc;
    }

    &__name {
        grid-column-start: 1;
        width: 98%;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__email {
        grid-column-start: 2;
        width: 98%;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__status {
        grid-column-start: 3;
        justify-self: center;

        select {
            padding: 4px 16px;
            border: none;
            border-radius: 16px;
            background-color: #f9fafc;
            color: @primary;
            font-size: 16px;
            cursor: pointer;
        
            &:focus {
              outline: none;
            }
          }
    }

    &__size {
        grid-column-start: 4;
        width: auto;
        display: flex;
        flex-direction: column;
        font-size: 12px;
        padding: 4px 20px;

        .progressBar {
            width: auto;
            padding: 4px 0;

            &__line, &__upload {
                height: 8px;
            }

            &__percent {
                display: none;
            }
        }
    }

    &__actions {
        grid-column-start: 5;

        .primaryButton {
            padding: 8px 24px;
        }
    }
}

@media only screen and (max-width: 660px){
    .user {
        grid-template-columns: 1.5fr 2fr 2fr;

        &__name {
            margin-left: 4px;
        }

        &__email {
            display: none
        }

        &__status {
            grid-column-start: 2;
        }
    
        &__size {
            grid-column-start: 3;
        }
    }
}