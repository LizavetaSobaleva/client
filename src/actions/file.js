import { hideLoader, showLoader } from "../reducers/appReducer"
import { addFile, deleteFileAction, setFiles, setStructure } from "../reducers/fileReducer"
import { addUploadFile, changeUploadFile, showUploader } from "../reducers/uploadReducer"
import { instanceAxios } from "./user"

export function getFiles(dirId) {
  return async (dispatch) => {
    try {
      dispatch(showLoader())
      const response = await instanceAxios.get(`files${dirId ? '?parent='+dirId : ''}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setFiles(response.data))
    } catch (e) {
      console.log(e.response.data.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}

export function createDir(dirId, name) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.post(`files`, {
          name,
          parent: dirId,
          type: 'dir'
        },
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
        )
      dispatch(addFile(response.data))
    } catch (e) {
      alert(e.response.data.message)
    }
  }
} 

export function uploadFile(file, dirId) {
  return async (dispatch) => {
    try {
      const formData = new FormData()
      formData.append('file', file)
      if(dirId) {
        formData.append('parent', dirId)
      }
      const uploadFile = {name: file.name, process: 0, id: Date.now()}
      dispatch(showUploader())
      dispatch(addUploadFile(uploadFile))
      const response = await instanceAxios.post(`files/upload`, formData, {
                headers: {Authorization: `Bearer ${localStorage.getItem('token')}`},
                onUploadProgress: progressEvent => {
            const totalLength = progressEvent.event.lengthComputable ? progressEvent.event.total : progressEvent.event.target.getResponseHeader('content-length') || progressEvent.event.target.getResponseHeader('x-decompressed-content-length')
            if(totalLength) {
              uploadFile.progress = Math.round((progressEvent.loaded * 100) / totalLength)
              dispatch(changeUploadFile(uploadFile))
            }
          }
        }
      )
      dispatch(addFile(response.data))
    } catch (e) {
      alert(e.response.data.message)
    }
  }
} 

export async function downloadFile(file) {
  try {
    const response = await instanceAxios.get(`files/download?id=${file._id}`, {
      responseType: 'blob',
      headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
    )
    const blob = new Blob([response.data], { type: response.headers['content-type']})

    const downloadUrl = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = downloadUrl
    link.download = file.name

    document.body.appendChild(link)
    link.click()
    link.remove()
  } catch (e) {
    console.log(e.response.data.message)
  }
}

export function deleteFile(file) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.delete(`files?id=${file._id}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(deleteFileAction(file._id))
      alert(response.data.message)
    } catch (e) {
      alert(e.response.data.message)
    }
  }
}

export function searchFiles(search) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.get(`files/search?search=${search}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setFiles(response.data))
    } catch (e) {
      alert(e?.response?.data?.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}

export function getStructure() {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.get(`files/structure`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setStructure(response.data))
    } catch (e) {
      alert(e?.response?.data?.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}