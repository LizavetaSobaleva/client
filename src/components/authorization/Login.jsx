import React, { useState } from "react";
import { Link } from 'react-router-dom';
import "./authorization.less";
import Input from "../UI/input/Input";
import PrimaryButton from "../UI/primaryButton/PrimaryButton";
import { useDispatch } from 'react-redux';
import { login } from "../../actions/user";

const Login = () => {
    const [email, setEmail] = useState("")
    const [password, setPassword] = useState("")
    const dispatch = useDispatch()


  return (
    <div className="authorization">
      <div className="authorization__container">
        <div className="authorization__label">
          <h1>Log in to cloud</h1>
        </div>
        <div className="authorization__form">
          <div className="authorization__item">
            <label>Email</label>
            <Input value={email} onChange={(event) => setEmail(event.target.value)} type="email" placeholder="Enter your email" />
          </div>

          <div className="authorization__item">
            <label>Password</label>
            <Input value={password} onChange={(event) => setPassword(event.target.value)} type="password" placeholder="Enter your password" />
          </div>

          <PrimaryButton onClick={() => dispatch(login(email, password))}>Log in</PrimaryButton>
        </div>
        <div className="authorization__login">
          <Link to="/registration">Create an account</Link>
        </div>
      </div>
    </div>
  );
};

export default Login;
