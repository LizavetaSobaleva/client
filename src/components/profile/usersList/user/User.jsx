import React, { useState } from "react";
import "./user.less";
import sizeFormat from "../../../../utils/sizeFormat";
import ProgressBar from "../../../UI/progressBar/ProgressBar";
import { useDispatch } from "react-redux";
import { changeUserStatus } from "../../../../actions/user";
import PrimaryButton from "../../../UI/primaryButton/PrimaryButton";
import SecondaryButton from "../../../UI/secondaryButton/SecondaryButton";
import Label from "../../../UI/label/Label";
import Selector from "../../../UI/selector/Selector";

const User = ({ user, fetchUsers }) => {
  const dispatch = useDispatch();
  const usedSpace = ((user.usedSpace / user.diskSpace) * 100).toFixed();
  const [status, setStatus] = useState(user.status);
  const [isEditing, setIsEditing] = useState(false);

  const handleStatusChange = (newStatus) => {
    setStatus(newStatus);
  };

  const handleSave = async () => {
    await dispatch(changeUserStatus(user._id, status));
    setIsEditing(false);
    fetchUsers();
  };

  const statusOptions = [
    { value: 'standard', label: 'standard' },
    { value: 'premium', label: 'premium' },
    { value: 'admin', label: 'admin' },
  ];

  return (
    <div className="user">
      <div className="user__name">{user.name}</div>
      <div className="user__email">{user.email}</div>

      <div className="user__status">
        {isEditing ? (
          <Selector value={status} options={statusOptions} onChange={handleStatusChange} />
        ) : (
          <Label status={user.status}>{user.status}</Label>
        )}
      </div>

      <div className="user__size">
        {sizeFormat(user.usedSpace)} from {sizeFormat(user.diskSpace)}
        <ProgressBar progress={usedSpace} />
      </div>

      <div className="user__actions">
        {isEditing ? (
          <PrimaryButton onClick={handleSave}>Save</PrimaryButton>
        ) : (
          <SecondaryButton onClick={() => setIsEditing(true)}>Edit</SecondaryButton>
        )}
      </div>
    </div>
  );
};

export default User;