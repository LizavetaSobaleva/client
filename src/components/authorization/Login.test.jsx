import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import Login from './Login';
import { login } from '../../actions/user';

jest.mock('../../actions/user', () => ({
  login: jest.fn().mockReturnValue({ type: 'SET_USER' }),
}));

const mockStore = configureStore([]);
const store = mockStore({});

describe('Login Component', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );
    store.dispatch = jest.fn();
  });

  test('renders Login form without crashing', () => {
    expect(screen.getByTestId('formTitle')).toBeInTheDocument();
    expect(screen.getByTestId('emailInput')).toBeInTheDocument();
    expect(screen.getByTestId('passwordInput')).toBeInTheDocument();
    expect(screen.getByTestId('loginButton')).toBeInTheDocument();
    expect(screen.getByTestId('registrationLink')).toBeInTheDocument();
  });

  test('allows user to input email', () => {
    const emailInput = screen.getByTestId('emailInput');
    fireEvent.change(emailInput, { target: { value: 'john.doe@example.com' } });
    expect(emailInput).toHaveValue('john.doe@example.com');
  });

  test('allows user to input password', () => {
    const passwordInput = screen.getByTestId('passwordInput');
    fireEvent.change(passwordInput, { target: { value: 'password123' } });
    expect(passwordInput).toHaveValue('password123');
  });

  test('login button triggers login action', () => {
    fireEvent.change(screen.getByTestId('emailInput'), { target: { value: 'john.doe@example.com' } });
    fireEvent.change(screen.getByTestId('passwordInput'), { target: { value: 'password123' } });

    fireEvent.click(screen.getByTestId('loginButton'));

    expect(store.dispatch).toHaveBeenCalledWith(login('john.doe@example.com', 'password123'));
  });
});
