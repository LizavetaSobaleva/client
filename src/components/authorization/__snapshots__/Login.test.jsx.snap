// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component renders Login form without crashing 1`] = `
<DocumentFragment>
  <div
    class="authorization"
  >
    <div
      class="authorization__container"
    >
      <div
        class="authorization__label"
        data-testid="formTitle"
      >
        <h1>
          Log in to cloud
        </h1>
      </div>
      <div
        class="authorization__form"
      >
        <div
          class="authorization__item"
        >
          <label
            for="emailInput"
          >
            Email
          </label>
          <input
            data-testid="emailInput"
            id="emailInput"
            placeholder="Enter your email"
            type="email"
            value=""
          />
        </div>
        <div
          class="authorization__item"
        >
          <label
            for="passwordInput"
          >
            Password
          </label>
          <input
            data-testid="passwordInput"
            id="passwordInput"
            placeholder="Enter your password"
            type="password"
            value=""
          />
        </div>
        <button
          class="primaryButton"
          data-testid="loginButton"
        >
          Log in
        </button>
      </div>
      <div
        class="authorization__login"
        data-testid="registrationLink"
      >
        <a
          href="/registration"
        >
          Create an account
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
